internal/fxlog/fxlog.go

```diff
@@ -21,86 +21,127 @@
+// by fx.Supply.
+type SupplyEvent struct {
+	Constructor interface{}
 }
 
-// F is a constructor for a Field.
-func F(key string, value interface{}) Field {
-	return Field{
-		Key:   key,
-		Value: value,
-	}
+// ProvideEvent is emitted whenever Provide was called and is not provided by fx.Supply.
+type ProvideEvent struct {
+	Constructor interface{}
 }
 
-// Field defines a field used inside an internal logging Entry.
-type Field struct {
-	Key   string
-	Value interface{}
+// InvokeEvent is emitted whenever a function is invoked.
+type InvokeEvent struct {
+	Function interface{}
 }
 
-// Logger defines interface used for logging.
-type Logger interface {
-	Log(entry Entry)
+// InvokeFailedEvent is emitted when fx.Invoke has failed.
+type InvokeFailedEvent struct {
+	Function interface{}
+	Err      error
+	Stack    fxreflect.Stack
 }
 
+// StartFailureError is emitted right before exiting after failing to start.
+type StartFailureError struct{ Err error }
+
+// StopSignalEvent is emitted whenever application receives a signal after
+// starting the application.
+type StopSignalEvent struct{ Signal string }
+
+// StopErrorEvent is emitted whenever we fail to stop cleanly.
+type StopErrorEvent struct{ Err error }
+
+// StartErrorEvent is emitted whenever a service fails to start.
+type StartErrorEvent struct{ Err error }
+
+// StartRollbackError is emitted whenever we fail to rollback cleanly after
+// a start error.
+type StartRollbackError struct{ Err error }
+
+// RunningEvent is emitted whenever an application is started successfully.
+type RunningEvent struct{}
+
 var _ Logger = (*zapLogger)(nil)
 
 type zapLogger struct {
 	logger *zap.Logger
 }
 
-func encodeFields(fields []Field, stack string) []zap.Field {
-	var fs []zap.Field
-	for _, field := range fields {
-		fs = append(fs, zap.Any(field.Key, field.Value))
-	}
-	if stack != "" {
-		fs = append(fs, zap.String("stack", stack))
-	}
-
-	return fs
-}
-
-func (l *zapLogger) Log(entry Entry) {
-	switch entry.Level {
-	case InfoLevel:
-		l.logger.Info(entry.Message, encodeFields(entry.Fields, entry.Stack)...)
-	case ErrorLevel:
-		l.logger.Error(entry.Message, encodeFields(entry.Fields, entry.Stack)...)
+func (l *zapLogger) LogEvent(event Event) {
+	switch e := event.(type) {
+	case LifecycleOnStartEvent:
+		l.logger.Info("starting", zap.String("caller", e.Caller))
+	case LifecycleOnStopEvent:
+		l.logger.Info("stopping", zap.String("caller", e.Caller))
+	case ApplyOptionsError:
+		l.logger.Error("error encountered while applying options", zap.Error(e.Err))
+	case SupplyEvent:
+		for _, rtype := range fxreflect.ReturnTypes(e.Constructor) {
+			l.logger.Info("supplying",
+				zap.String("constructor", fxreflect.FuncName(e.Constructor)),
+				zap.String("type", rtype),
+			)
+		}
+	case ProvideEvent:
+		for _, rtype := range fxreflect.ReturnTypes(e.Constructor) {
+			l.logger.Info("providing",
+				zap.String("constructor", fxreflect.FuncName(e.Constructor)),
```

Comment:
Not a blocker for this diff but FYI: once https://github.com/uber-go/dig/pull/280 lands, we might be able to get rid of fxreflect.FuncName completely.